Add basic scrapper

thinkjin99 opened the pull request:
Basic scrapper scraps information of latest articles in blind.
Scrapper acts async. Only works in local now.
test.csv file is needed to run. test.csv is local version of main csv

thinkjin99 committed:
Add basic scrapper

Basic scrapper scrpas information of latest articles in blind.
Scrapper acts async. Only works in local now.
test.csv file is needed to run. test.csv is local version of main csv

kkanggu commented:
`main` branch는 배포용, 개발 환경은 `dev` 입니다.

kkanggu commented:
본문이 영어라 빠르게 읽히지 않네요. 이 부분 고민하여 Github Guide에 적용 혹은 미팅때 논의 사항으로 가져오겠습니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,209 @@
+    
+    Return:
+        list: links of valid articles, not exist in csv file
+    """
```
주석 없이 바로 읽히고 이해가 되는 코드가 베스트이고, 그렇지 않을 경우 주석 사용이 권장됩니다.
함수명을 통해 무슨 작업을 하고 뭘 가져오는지 바로 알 수 있으나, 파라미터는 직관적으로 알아차리기 어렵네요. 더 좋은 네이밍이 없을까요?

thinkjin99 replied:
해당 파라미터의 경우 이미  csv 파일에 저장된 게시글들의 고유 코드 값을 의미합니다. 말씀해주신 대로 csv_code는 난해한 면이 존재하므로 이미 저장된 값이라는 점을 강조해 exist_ariticle_codes 로 수정하는거 제안드립니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,209 @@
+    url = "https://www.teamblind.com/kr/topics/%EB%B8%94%EB%9D%BC%EB%B8%94%EB%9D%BC"
+    
+    resp = requests.get(url, headers=headers)
+    soup = bs(resp.text, 'html.parser')
```
requests.get 실패시 프로그램이 뻗어버리지는 않나요? 예외 처리가 보이지 않아 불안합니다.

thinkjin99 replied:
이 부분은 제가 부주의했습니다. 예외처리하는 편이 좋아보입니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,209 @@
+        {date:"2022-08-01-07:00", "name": "한샘", title:"블라보면 징징이가 왜이리 많은지", content:"9급이나 대기업이나. 걍 하면 되는거지 꼭 물고 뜯고 뒤지게 싸우는지 왜", "url":"https://..."},
+        {date:"2022-09-01-07:00", "name": "공무원", title:"번개하실분", content:"성수동 번개 구합니다.", "url":"https://..."}, ...
+        ]
+    """
```
위와 동일합니다.
동일한 dict 포맷을 사용하는 것으로 보이는데, 여러 함수에서 주석으로 알려주는 것 보다는 개별 문서로 확인하는 것이 더 나을 수도 있을 것 같습니다.

kkanggu replied:
추가로, 파라미터가 무슨 link인지 더 명확히 알려준다면 좋을 것 같습니다.

thinkjin99 replied:
links는 각각의 게시글에 대한 url을 의미합니다. 명확한 의미를 전달하기 위해서 aritcle_links나 article_urls로 수정하는 편이 적절해 보입니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,209 @@
+
+    Returns:
+        dictionary: {date:"2022-08-01-07:00", "name": "한샘", title:"블라보면 징징이가 왜이리 많은지", content:"9급이나 대기업이나. 걍 하면 되는거지 꼭 물고 뜯고 뒤지게 싸우는지 왜", "url":"https://..."},
+    """
```
위와 동일합니다.
함수명에서 link와 links를 발견하지 못하면 단일 링크인지 여러 링크인지 바로 알아차리기 어려운 것 같습니다.
info가 정확히 무슨 정보인지 코드에서는 바로 알 수 있는데, 주석에도 명시하면 좋은지 다른 분들의 의견도 들어보면 좋을 것 같습니다.

thinkjin99 replied:
info를 한번 정도는 주석에서 설명해줄 필요가 있을 것 같습니다. info에 대해서 한번만 주석에서 설명하고 이를 활용한 infos와 같은 변수에 대한 설명은  유추가 가능하므로  주석에서 제거하는게 좋아 보입니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,209 @@
+    except Exception as e:
+        print(">> Error occured in url")
+        print(resp.url)
+        print(e)
```
time에 관한 정보는 `e`에 포함이 되어있나요?

thinkjin99 replied:
e는 에러사항만 포함하고 있어 time에 관한 정보는 없습니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,209 @@
+    except Exception as e:
+        print(">> Convert date has problem")
+        print(f"amonut :{time_amount} unit: {time_unit}")
+        print(e)
```
파라미터로 넘어온 raw_date에 대한 정보도 같이 나오면 좋을 것 같습니다.

thinkjin99 replied:
반영하겠습니다.

thinkjin99 committed:
Rename variable

Replaces word "link" contains in variable name to "article_url".
link -> article_url, links -> article_urls.
Remove description about infos
Additionally reflects small issues commented in pull request #1

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,220 @@
+    
+    except Exception as e:
+        print(">> Error occured in get article urls")
+        print(e)
```
예외 발생 시각도 기록되어야 할 것 같습니다.

choiking10 replied:
logging 모듈을 사용해서 로깅해보는 것을 추천합니다.  
https://minimin2.tistory.com/41  참고해주세요

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,220 @@
+    except Exception as e:
+        print(">> Error occured in get article info")
+        print(resp.url)
+        print(e)
```
위와 동일하게 시각도 표기되면 좋을 것 같습니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,220 @@
+            elif time_unit == "일":
+                time_ago = datetime.timedelta(days=time_amount)
+        
+        else:
+            if time_unit ==  "방금":
+                time_ago = datetime.timedelta(minutes=0)
+            elif time_unit == "어제":
+                time_ago = datetime.timedelta(days=1)
```
정확한 시각 대신 방금, 어제와 같이 추상적인 시각으로만 가져올 수 있나요?

thinkjin99 replied:
블라블라 게시판의 시각을 표기하는 방식이 깔끔한 datetime형식이 아니라. 방금,1분전,1시간전,하루전, 2일전 같은 추상적인 형태로만 전달됩니다. 정확하게 가져오려면 네트워크 부분의 response를 뜯어봐야 할 것 같은데.. 찾기가 쉽지가 않습니다.

choiking10 commented on scrap_blind.py:
```diff
@@ -0,0 +1,220 @@
+        list: urls of valid articles, not exist in csv file
+    """
+    try:
+        url = "https://www.teamblind.com/kr/topics/%EB%B8%94%EB%9D%BC%EB%B8%94%EB%9D%BC"
```
url encode를 따로 적용해서 가독성을 높이는게 좋을 것 같습니다. 
제가 분석한 건 아니지만, 보니까 topic에 따라 뒤에 오는 path param이달라질 것 같은데, 이부분을 변수로 빼두고 확장성을 고려해주시면 좋을 것 같아요.

thinkjin99 replied:
topic 이후에 오는 param은 게시판 이름을 뜻합니다. 현재는 블라블라만을 대상으로해 저런 방식으로 구축 했지만, 지적해주신 것 처럼 확장성을 고려한다면 변수로 처리하는 것이 맞을 것 같습니다.

thinkjin99 replied:
처리해서 반영했습니다.

choiking10 commented on scrap_blind.py:
```diff
@@ -0,0 +1,220 @@
+    """
+    try:
+        url = "https://www.teamblind.com/kr/topics/%EB%B8%94%EB%9D%BC%EB%B8%94%EB%9D%BC"
+        resp = requests.get(url, headers=headers)
```
asyncio에서 requests를 쓸 경우 제대로 async하게 동작하지 않을 것 같은데 확인 부탁드립니다.

thinkjin99 replied:
저 경우 async하게 동작하지 않는 것이 맞습니다. 현재 스크래핑은 메인 페이지에서 세부 페이지의 url을 스크래핑한 후 각 세부 페이지의 정보를 async하게 스크래핑 하는 방법으로 동작을 하고 있습니다. 위의 코드 부분은 메인 페이지에서 세부 페이지 들의 url만을 가져오는 리퀘스트여서 속도를 높일 필요가 없다 생각해 async하게 구성하지 않았습니다.

thinkjin99 replied:
메인페이지에 리퀘스트를 통해 게시글 별 URL 수집(sync) -> 각 URL 별 정보 스크래핑(async) 의 방식으로 동작한다고 보시면 됩니다.

choiking10 commented on scrap_blind.py:
```diff
@@ -0,0 +1,220 @@
+        print(date)
+        print(name)
+        print(content)
+        print(e)
```
logging

thinkjin99 replied:
로깅 레벨이나 로깅 내용 같은 경우 어떻게 처리해야할지 고민입니다. 우선 공식문서 따라서 간단한 로깅으로 구현했습니다

choiking10 commented on scrap_blind.py:
```diff
@@ -0,0 +1,220 @@
+    except Exception as e:
+        print(">> Convert date has problem")
+        print(f"raw_date: {raw_date} ,amonut :{time_amount} unit: {time_unit}")
+        print(e)
```
logging

thinkjin99 committed:
Add logger

Add logger. Logger prints log to console and crawl.log file.
crawl_logger set and create logger.

thinkjin99 committed:
Rename logger

thinkjin99 committed:
Refactor scrapping result

스크래핑 결과를 기존 csv에 저장하는 방식에서
json으로 저장하는 방식으로 변경했습니다.

thinkjin99 committed:
Add S3 upload script

스크래핑의 결과로 저장된 json 파일을
S3의 버킷에 업로드하는 스크립트 입니다.

kkanggu commented on scrap_blind.py:
```diff
@@ -0,0 +1,203 @@
+        return hrefs
+    
+    except Exception as e:
+        logger.error(e)
```
`my_logger.py`의 logger를 이용하여 미리 명시한 포맷에 맞게 error, 혹은 warning이 기록이 되는 방식일까요?

thinkjin99 replied:
맞습니다. my logger에 명시 해놓은 포맷에 따라 에러가 .log 파일에 작성되는 방식입니다.

kkanggu commented on upload_json.py:
```diff
@@ -0,0 +1,45 @@
+    logger = my_logger.create_logger()
+    #AWS is utc, local file is in seoul time zone
+    # AWS_TIMEGAP = datetime.timedelta(hours=9)
+    try:
+        s3 = boto3.client('s3', region_name='ap-northeast-2')
+        bucket_name = "issue-keyword-storage"
+        bucket = s3.list_objects_v2(Bucket=bucket_name, Prefix= JSON_PATH, Delimiter='/')
+    
+    except Exception as e:
+        logger.error(e)
+
+    json_files = os.listdir(JSON_PATH)
+    file_name = json_files[-1]
+    
+    last_update_date = get_last_modified_date(bucket)
+    file_date = get_last_file_date(file_name)
+
+    if last_update_date < file_date:
+        file_path = f"{JSON_PATH}/{file_name}"
+        key = "json/" + file_name
+        s3.upload_file(file_path, bucket_name, key)
+    
+    else:
+        print("No need to update")
```
exception 발생한다면 s3에 문제가 생기는 것 같은데, 아래에 s3가 사용이 되네요.
위의 경우 프로그램이 비정상적으로 종료되지는 않나요?

thinkjin99 replied:
확인했습니다. try문 안에 포함시키는 방향으로 변경 진행했습니다.

kkanggu reviewed (commented):
고생하셨습니다.

thinkjin99 committed:
Refactor upload script

업로드 하는 방식을 함수로 묶어서 스크래핑 코드에서 한번에 실행하는 구조로 변경.
스크래핑 -> 파일 저장 -> 파일 읽기 -> 업로드 에서
스크래핑 -> 파일 저장 -> 업로드의 순서로 동작하게 구성.

thinkjin99 committed:
Add duplicate check and upload

가장 최근의 json 파일과 비교해 중복되는 게시글은 스크래핑하지 않는다.
스크래핑 직후 업로드를 진행한다.
주석을 약간 변경하고 추가했다.

thinkjin99 committed:
Refacotr path option

pathlib를 활용해 스크립트가 참조하는 파일들의 위치를 절대 경로에서
스크립트가 실행되는 폴더를 기준으로 변경하였습니다.
서버 업로드 후 파일 참조가 적절히 이뤄지지 않아 문제가 발생했었음.

thinkjin99 committed:
Set log file path

기존의 로그 파일 경로를 변경하고 상대경로로 설정함
로그 파일은 이제 log 폴더 안에 scrap.log로 생성

kkanggu reviewed (approved):
자체 판단하에 merge
